core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -292,10 +330,20 @@ private static void validate(SnapshotId snapshotId) {
      * @param partial                    allow partial snapshots
      * @param userCreateSnapshotListener listener
      */
-    private void beginSnapshot(ClusterState clusterState, final SnapshotsInProgress.Entry snapshot, final boolean partial, final CreateSnapshotListener userCreateSnapshotListener) {
+    private void beginSnapshot(final ClusterState clusterState,
+                               final SnapshotsInProgress.Entry snapshot,
+                               final boolean partial,
+                               final CreateSnapshotListener userCreateSnapshotListener) {
         boolean snapshotCreated = false;
         try {
-            Repository repository = repositoriesService.repository(snapshot.snapshotId().getRepository());
+            // check if the snapshot name already exists in the repository
+            if (snapshotNameExists(snapshot.snapshot())) {
```

Comment:
There are a few issues here. First I think we have a race condition here. Moving this check to the the repository level will not remove it right now, but will tighten the loop. Hopefully, we will solve it there at some point of time with tightening the blob store api and adding locks. The second issue is that it can be done more effectively on the repository level.